File: jdroid-java-core/src/main/java/com/jdroid/java/concurrent/ExecutorUtils.java
package com.jdroid.java.concurrent;

import com.jdroid.java.utils.LoggerUtils;

import org.slf4j.Logger;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

public final class ExecutorUtils {
	
	private static final Logger LOGGER = LoggerUtils.getLogger(ExecutorUtils.class);
	
	// Default amount of thread inside the pool
	private static final int DEFAULT_THREAD_POOL_SIZE = 5;
	
	private static final Executor fixedExecutor = Executors.newCachedThreadPool(new NormalPriorityThreadFactory());
	
	private static final Executor fixedLowPriorityExecutor = Executors.newFixedThreadPool(DEFAULT_THREAD_POOL_SIZE,
		new LowPriorityThreadFactory());
	
	private static final ScheduledExecutorService scheduledExecutor = Executors.newScheduledThreadPool(1,
		new NormalPriorityThreadFactory("schedule"));
	
	/**
	 * @param runnable The {@link Runnable} task
	 */
	public static void execute(Runnable runnable) {
		fixedExecutor.execute(runnable);
	}
	
	public static void executeWithLowPriority(Runnable runnable) {
		fixedLowPriorityExecutor.execute(runnable);
	}
	
	public static void schedule(Runnable runnable, Long delay, TimeUnit timeUnit) {
		scheduledExecutor.schedule(runnable, delay, timeUnit);
	}
	
	public static ScheduledFuture<?> schedule(Runnable runnable, Long delay, Long period, TimeUnit timeUnit) {
		return scheduledExecutor.scheduleAtFixedRate(runnable, delay, period, timeUnit);
	}
	
	public static void sleep(int delay, TimeUnit timeUnit) {
		try {
			Thread.sleep(timeUnit.toMillis(delay));
		} catch (InterruptedException e) {
			LoggerUtils.logHandledException(LOGGER, e);
		}
	}
}

File: jdroid-java-core/src/main/java/com/jdroid/java/concurrent/ExecutorUtils.kt
package com.jdroid.java.concurrent

import com.jdroid.java.utils.LoggerUtils
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

object ExecutorUtils {

    private val LOGGER = LoggerUtils.getLogger(ExecutorUtils::class.java)

    // Default amount of thread inside the pool
    private const val DEFAULT_THREAD_POOL_SIZE = 5

    private val fixedExecutor = Executors.newCachedThreadPool(NormalPriorityThreadFactory())

    private val fixedLowPriorityExecutor = Executors.newFixedThreadPool(DEFAULT_THREAD_POOL_SIZE, LowPriorityThreadFactory())

    private val scheduledExecutor = Executors.newScheduledThreadPool(1, NormalPriorityThreadFactory("schedule"))

    /**
     * @param runnable The [Runnable] task
     */
    fun execute(runnable: Runnable) {
        fixedExecutor.execute(runnable)
    }

    fun executeWithLowPriority(runnable: Runnable) {
        fixedLowPriorityExecutor.execute(runnable)
    }

    fun schedule(runnable: Runnable, delay: Long?, timeUnit: TimeUnit) {
        scheduledExecutor.schedule(runnable, delay!!, timeUnit)
    }

    fun schedule(runnable: Runnable, delay: Long?, period: Long?, timeUnit: TimeUnit): ScheduledFuture<*> {
        return scheduledExecutor.scheduleAtFixedRate(runnable, delay!!, period!!, timeUnit)
    }

    fun sleep(delay: Int, timeUnit: TimeUnit) {
        try {
            Thread.sleep(timeUnit.toMillis(delay.toLong()))
        } catch (e: InterruptedException) {
            LoggerUtils.logHandledException(LOGGER, e)
        }
    }
}
